syntax = "proto3";

option go_package = "HydraSAT/src/master/proto";

service SolverService {
    // Worker tell the Master its ID
    // Master then sends CountRequest to the worker 
    rpc GetTask (RegisterRequest) returns (CountRequest);

    // When the worker finish solving 
    // it sends the result back to the Master using this rpc
    rpc SubmitResult (CountResponse) returns (Empty);
}
message RegisterRequest {
    string worker_id = 1; // ID of the worker
}

message Empty {}
message CountRequest {
    int64 task_id = 1; // unique ID for the counting task
    int32 num_vars = 2; // number of variables in the CNF
    int32 num_clauses = 3; // number of clauses in the CNF
    string formula_body = 4; // raw DIMACS body of the CNF (without the header)
    repeated int32 literals = 5; // the cube
    int32 timeout_sec = 6;  // timeout in seconds for the solver to find a solution
}

message CountResponse {
    string count = 1; // the number of solutions as a string to handle large numbers
    double duration_sec = 2; // time taken to solve in seconds
    bool timed_out = 3; // indicates if the solver timed out
    int64 task_id = 4; // the ID of the task (for tracking purposes)
    string worker_id = 5; // the ID of the worker that processed this task
}
